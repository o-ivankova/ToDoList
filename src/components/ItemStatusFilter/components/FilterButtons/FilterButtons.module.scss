%filterButton-shared {
    background: none;
    font-size: 14px;
    border: none;
    border-style: hidden;
}

.filterButton {
    @extend %filterButton-shared;
    color: grey;
}

.filterButton:hover {
    @extend %filterButton-shared;
    color: #363636;
}

.filterButtonActive {
    @extend %filterButton-shared;
   color: #74269b;
}