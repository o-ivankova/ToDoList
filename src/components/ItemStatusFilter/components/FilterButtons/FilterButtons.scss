.filter-button {
    border-style: hidden;
    background: none;
    color: grey;
    font-size: 14px;
}

.filter-button:hover {
    color: #363636;
}

.filter-button.active {
    border: none;
    border-style: hidden;
    color: #74269b;
}