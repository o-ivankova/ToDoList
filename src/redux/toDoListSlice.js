import { createSlice } from "@reduxjs/toolkit";
import storage from "redux-persist/lib/storage";
import { persistReducer } from "redux-persist";

const initialState = {
  elements: [],
  elementId: 0,
  filter: "all",
};

export const toDoListSlice = createSlice({
  name: "toDoList",
  initialState,
  reducers: {
    addElement: (state, action) => {
      const input = action.payload;

      if (input !== "") {
        state.elementId += 1;

        const newElement = {
          content: input,
          completed: false,
          id: state.elementId,
        };

        state.elements.push(newElement);
      }
    },

    deleteElement: (state, action) => {
      const id = action.payload;
      state.elements = state.elements.filter((el) => el.id !== id);
    },

    deleteAllCompleted: (state) => {
      state.elements = state.elements.filter((el) => !el.completed);
    },

    checkCompleted: (state, action) => {
      const id = action.payload;
      const index = state.elements.findIndex((el) => el.id === id);
      state.elements[index].completed = !state.elements[index].completed;
    },

    changeFilter: (state, action) => {
      const filter = action.payload;
      state.filter = filter;
    },

    reorderElements: (state, action) => {
      const result = action.payload;
      if (!result.destination) return;

      const [reorderedEl] = state.elements.splice(result.source.index, 1);
      state.elements.splice(result.destination.index, 0, reorderedEl);
    },
  },
});

export const {
  addElement,
  deleteElement,
  checkCompleted,
  deleteAllCompleted,
  changeFilter,
  reorderElements,
} = toDoListSlice.actions;

const persistConfig = {
  key: "root",
  storage,
};

const { reducer } = toDoListSlice;

export const persistedReducer = persistReducer(persistConfig, reducer);

// export default reducer;
