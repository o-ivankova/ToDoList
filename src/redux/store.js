import { configureStore } from "@reduxjs/toolkit";
import {persistedReducer} from "./toDoListSlice";
import { persistStore } from 'redux-persist'

export const store = configureStore({
  reducer: {
    toDoList: persistedReducer,
  },
});

export const persistor = persistStore(store);