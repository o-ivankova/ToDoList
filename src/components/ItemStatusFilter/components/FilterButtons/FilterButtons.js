import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { changeFilter } from "../../../../redux/toDoListSlice";
import "./FilterButtons.scss";

const FilterButtons = () => {
  const filter = useSelector((state) => state.toDoList.filter);
  const dispatch = useDispatch();

  const onFilterChange = (filter) => {
    dispatch(changeFilter(filter));
  };

  const buttonsData = [
    { name: "all", label: "All" },
    { name: "active", label: "Active" },
    { name: "completed", label: "Completed" },
  ];

  const buttons = buttonsData.map(({ name, label }) => {
    const isActive = filter === name;
    const clazz = isActive ? "active" : "";
    return (
      <button
        type="button"
        className={`filter-button ${clazz}`}
        key={name}
        onClick={() => onFilterChange(name)}
      >
        {label}
      </button>
    );
  });

  return <div className="filter-buttons">{buttons}</div>;
};

export default FilterButtons;
