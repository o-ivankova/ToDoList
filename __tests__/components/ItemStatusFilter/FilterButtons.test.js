import React from "react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import { expect, test, describe, jest, afterAll } from "@jest/globals";
import { fireEvent, render, screen } from "@testing-library/react";
import FilterButtons from "../../../src/components/ItemStatusFilter/components/FilterButtons/FilterButtons";
import { setupStore } from "../../../src/redux/store";
import renderer from "react-test-renderer";

describe("FilterButtons tests", () => {
  test("component is rendered", () => {
    const preloadedState = {
      toDoList: {
        filter: "all",
      },
    };

    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <FilterButtons />
      </Provider>
    );

    expect(screen.getByTestId("filter-buttons")).toBeInTheDocument();
  });

  test("all 3 buttons are rendered", () => {
    const preloadedState = {
      toDoList: {
        filter: "all",
      },
    };

    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <FilterButtons />
      </Provider>
    );

    expect(screen.getByText("All")).toBeInTheDocument();
    expect(screen.getByText("Active")).toBeInTheDocument();
    expect(screen.getByText("Completed")).toBeInTheDocument();
  });

  test("the completed button is active from the start", () => {
    const preloadedState = {
      toDoList: {
        filter: "completed",
      },
    };

    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <FilterButtons />
      </Provider>
    );

    expect(screen.getByText("Completed")).toHaveClass("filterButtonActive");
  });

  test("the active button is clicked", () => {
    const preloadedState = {
      toDoList: {
        filter: "all",
      },
    };

    const mockStore = setupStore(preloadedState);
    localStorage.setItem(
      "filter",
      JSON.stringify(preloadedState.toDoList.filter)
    );

    render(
      <Provider store={mockStore}>
        <FilterButtons />
      </Provider>
    );
    fireEvent.click(screen.getByText("Active"));

    const expectedFilter = "active";
    const filter = JSON.parse(localStorage.getItem("filter"));
    expect(filter).toEqual(expectedFilter);
  });

  test("snapshot test", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: true },
          { id: 2, content: "test2", completed: true },
          { id: 3, content: "test3", completed: false },
        ],
      },
    };
    const mockStore = setupStore(preloadedState);

    const domTree = renderer
      .create(
        <Provider store={mockStore}>
          <FilterButtons />
        </Provider>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  afterAll(() => {
    jest.clearAllMocks();
  });
});
