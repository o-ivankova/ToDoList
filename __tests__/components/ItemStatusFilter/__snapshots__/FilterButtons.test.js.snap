// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterButtons tests snapshot test 1`] = `
<div
  data-testid="filter-buttons"
>
  <button
    className="filterButton"
    onClick={[Function]}
    type="button"
  >
    All
  </button>
  <button
    className="filterButton"
    onClick={[Function]}
    type="button"
  >
    Active
  </button>
  <button
    className="filterButton"
    onClick={[Function]}
    type="button"
  >
    Completed
  </button>
</div>
`;
